metaseq/data/iterators.py

```diff
@@ -260,29 +265,62 @@ def state_dict(self):
             # small optimization: we advance the epoch before saving, so that
             # when loading later we don't end up fast-forwarding the iterator
             epoch = self.epoch + 1
-            iter_in_epoch = 0
+            sentences_consumed = [0 for _ in range(self.num_workers)]
+            n = 0
         else:
             epoch = self.epoch
-            iter_in_epoch = self.iterations_in_epoch
+            sentences_consumed = self._itr.sentences_consumed
+            n = self._itr.n
+
+        dataset = self.dataset
+        while not isinstance(dataset, DeferredDataset):
+            dataset = dataset.dataset
+        logger.info(f"Saving state_dict so we can skip workers quickly: {len(dataset.len_cache)} "
+                    f"entries in tokenization_cache, {sentences_consumed} sentences consumed per worker, iteration {n}")
         return {
             "epoch": epoch,
-            "iterations_in_epoch": iter_in_epoch,
+            "sentences_consumed": sentences_consumed,
+            "tokenization_cache": dataset.len_cache,
+            "n": n,
         }
 
     def load_state_dict(self, state_dict):
         """Copies the state of the iterator from the given *state_dict*."""
         self.epoch = state_dict["epoch"]
         if hasattr(self.dataset, "set_epoch"):
             self.dataset.set_epoch(self.epoch)
-        self._itr = self._get_iterator_for_epoch(self.epoch)
-        itr_pos = state_dict.get("iterations_in_epoch", 0)
-        if itr_pos > 0:
-            # fast-forward epoch iterator
-            logger.info(f"Fast-forwarding dataloader by {itr_pos} batches...")
-            t0 = time.time()
-            next(itertools.islice(self._itr, itr_pos, itr_pos), None)
-            t1 = time.time()
-            logger.info(f"done fast-forwarding dataloader in {t1 - t0:.1f} seconds")
+
+        # must be set before _get_iterator_for_epoch otherwise the datasets in the workers
+        # will not be copied with the right state
+        if "sentences_consumed" in state_dict and max(state_dict["sentences_consumed"]) > 0:
+            sentences_consumed = state_dict["sentences_consumed"]
+            logger.info(f"Skipping {sentences_consumed} sentences in each worker...")
+            num_workers = 1 if self.num_workers == 0 else self.num_workers
+            assert len(sentences_consumed) == num_workers,\
+                "changing the number of workers in the middle of a shard changes the order the data will be loaded in"
+            dataset = self.dataset
+            while not isinstance(dataset, SkipDeferredDataset):
+                dataset = dataset.dataset
```

Comment:
Agreed. I think in patch 2, I am going to combine  ShuffleBuffer-GroupDocumentContentIntoBatches-DocumentShuffle into a single dataset object. This (1) improves fast-forward performance a lot by cutting down on byte codes executed, but also (2) means that the skip and defer logic can just be built into this single iterator rather than having to be sandwiched by these extra objects.